handlers/product-switch-api/src/frequencySwitchEndpoint.ts

```diff
@@ -0,0 +1,745 @@
+): { ratePlan: RatePlan; charge: RatePlanCharge } {
+	// Check subscription status if account is provided
+	assertValidState(
+		subscription.status === 'Active',
+		frequencySwitchValidationRequirements.subscriptionActive,
+		subscription.status,
+	);
+
+	// Check for outstanding unpaid invoices if account is provided
+	assertValidState(
+		account.metrics.totalInvoiceBalance === 0,
+		frequencySwitchValidationRequirements.zeroAccountBalance,
+		`${account.metrics.totalInvoiceBalance} ${account.metrics.currency}`,
+	);
+
+	// Log diagnostic info about which charges are being filtered
+	const initialCharges = subscription.ratePlans
+		.filter((rp) => rp.lastChangeType !== 'Remove')
+		.flatMap((rp) =>
+			rp.ratePlanCharges.map((c) => ({ ratePlan: rp, charge: c })),
+		);
+
+	logger.log(
+		`Found ${initialCharges.length} rate plan charges before filtering`,
+	);
+
+	const candidateCharges = initialCharges
+		.filter(({ charge }) => {
+			if (charge.name !== 'Subscription') {
+				logger.log(
+					`Filtering out charge ${charge.id}: name is "${charge.name}", not "Subscription"`,
+				);
+				return false;
+			}
+			return true;
+		})
+		.filter(({ charge }) => {
+			if (charge.type !== 'Recurring') {
+				logger.log(
+					`Filtering out charge ${charge.id}: type is "${charge.type}", not "Recurring"`,
+				);
+				return false;
+			}
+			return true;
+		})
+		.filter(({ charge }) => {
+			if (charge.effectiveStartDate > today) {
+				logger.log(
+					`Filtering out charge ${charge.id}: effectiveStartDate ${charge.effectiveStartDate.toISOString()} is in the future`,
+				);
+				return false;
+			}
+			if (charge.effectiveEndDate <= today) {
+				logger.log(
+					`Filtering out charge ${charge.id}: effectiveEndDate ${charge.effectiveEndDate.toISOString()} is in the past`,
+				);
+				return false;
+			}
+			return true;
+		})
+		.filter(({ charge }) => {
+			if (charge.chargedThroughDate && charge.chargedThroughDate < today) {
+				logger.log(
+					`Filtering out charge ${charge.id}: chargedThroughDate ${charge.chargedThroughDate.toISOString()} is in the past`,
+				);
+				return false;
+			}
+			return true;
+		})
+		.filter(({ charge }) => {
+			if (
+				charge.billingPeriod !== 'Month' &&
+				charge.billingPeriod !== 'Annual'
+			) {
+				logger.log(
+					`Filtering out charge ${charge.id}: billingPeriod is "${charge.billingPeriod}", not "Month" or "Annual"`,
+				);
+				return false;
+			}
+			return true;
+		});
+
+	if (candidateCharges.length === 0) {
+		logger.log(
+			`No eligible charges found after filtering. Initial count: ${initialCharges.length}`,
+		);
+	}
+
+	assertValidState(
+		candidateCharges.length > 0,
+		frequencySwitchValidationRequirements.hasEligibleCharges,
+		`${candidateCharges.length} charges found`,
+	);
+
+	assertValidState(
+		candidateCharges.length === 1,
+		frequencySwitchValidationRequirements.singleEligibleCharge,
+		`${candidateCharges.length} charges found`,
+	);
+
+	return candidateCharges[0]!;
```

Comment:
if you change the condition slightly, the type checker doesn't need the `!`
```suggestion
	assertValidState(
		candidateCharges[0] !== undefined,
		frequencySwitchValidationRequirements.hasEligibleCharges,
		`${candidateCharges.length} charges found`,
	);

	assertValidState(
		candidateCharges.length === 1,
		frequencySwitchValidationRequirements.singleEligibleCharge,
		`${candidateCharges.length} charges found`,
	);

	return candidateCharges[0];
```

Reply:
Or use the `isNotEmpty` function from the `arrayFunctions` module.

Reply:
Thank you for the suggestion @rupertbates 🙏